### PR TITLE
Users crud changes

### PR DESCRIPTION
**Pull request by carlosbflores**
Changes for creating users, setting roles, applications, and organizations. 
Added/changed controller endpoints, services, and repositories.

### CODE REVIEW

File: src/main/java/py/org/fundacionparaguaya/pspserver/web/rest/UserController.java [hunk 1]
```diff
@@ -55,9 +65,19 @@ public UserController(UserService userService) {
...
+	public ResponseEntity<PaginableList<UserDTO>>  getPaginatedUsers(@RequestParam(value = "page", required = false, defaultValue = "1") int page,
+			@RequestParam(value = "per_page", required = false, defaultValue = "12") int perPage,
+			@RequestParam(value = "sort_by", required = false, defaultValue = "username") String sortBy,
+			@RequestParam(value = "order", required = false, defaultValue = "asc") String orderBy){
+	//public ResponseEntity<List<UserDTO>> getAllUsers() {
+		PageRequest pageRequest = new PspPageRequest(page, perPage, orderBy, sortBy);
+		Page<UserDTO> pageProperties = userService.listUsers(pageRequest);
+		PaginableList<UserDTO> response = new PaginableList<>(pageProperties, pageProperties.getContent());
+		return ResponseEntity.ok(response);
+		/*
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/web/rest/UserController.java]: Remove commented code.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -60,7 +93,60 @@ public UserDTO getUserById(Long userId) {
...
+	public UserDTO addUserWithRoleAndApplication(UserRoleApplicationDTO userRoleApplicationDTO, UserDetailsDTO userDetails) {
+		// Create User
+		UserEntity user = new UserEntity();
+		BeanUtils.copyProperties(userRoleApplicationDTO.getUser(), user);
+		user.setPass(new BCryptPasswordEncoder().encode(user.getPass()));
+		user.setActive(true);
+		UserEntity newUser = userRepository.save(user);
+
+		if (userRoleApplicationDTO.getRole() != null) {
+			// Create UserRole
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: Let's put each of these if block of code inside a separate method.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -60,7 +93,60 @@ public UserDTO getUserById(Long userId) {
...
+			userRoleRepository.save(userRole);
+		}
+
+		if (userRoleApplicationDTO.getApplication() != null) {
+			// Create User-Application
+			UserApplicationEntity userApplicationEntity = new UserApplicationEntity();
+			userApplicationEntity.setUser(newUser);
+			ApplicationEntity application = applicationRepository.findById(userRoleApplicationDTO.getApplication().getId());
+			userApplicationEntity.setApplication(application);
+			userApplicationRepository.save(userApplicationEntity);
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: maybe we can put the content of this if inside a new method.

**Comment by rodrigojv** [comment]
@carlosbflores please run `mvn clean package` to see the errors.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/UserRoleService.java [hunk 4]
```diff
@@ -15,7 +15,9 @@
 	UserRoleDTO getUserRoleById(Long userRoleId);
 	
 	List<UserRoleDTO> getAllUserRoles();
-	
+
+	List<UserRoleDTO> getRolesByLoggedUser();
```
[the remark this answers is not in the document]
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/UserRoleService.java]: Changed method name to getAssignableRolesByUser() and added UserDetails as parameter

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java [hunk 5]
```diff
@@ -69,6 +73,27 @@ public UserRoleDTO getUserRoleById(Long userRoleId) {
 		return userRoleMapper.entityListToDtoList(userRole);
 	}
 
+	@Override
+	public List<UserRoleDTO> getRolesByLoggedUser() {
+		List<UserRoleDTO> rolesByLoggedUser = new ArrayList<UserRoleDTO>();
+
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
```
[the remark this answers is not in the document]
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: UserDetails passed as parameter

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java [hunk 6]
```diff
@@ -69,6 +73,27 @@ public UserRoleDTO getUserRoleById(Long userRoleId) {
...
+		} else if (authentication.getAuthorities().toArray()[0].toString().equals(Role.ROLE_HUB_ADMIN.name())) {
+			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_APP_ADMIN).build());
+			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_USER).build());
+		} else if (authentication.getAuthorities().toArray()[0].toString().equals(Role.ROLE_APP_ADMIN.name())) {
+			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_USER).build());
+			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_SURVEY_USER).build());
+//			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_SOCIAL_ASSISTANT).build());
+		}
+
+		return rolesByLoggedUser;
```
[the remark this answers is not in the document]
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: Changed method name to getAssignableRolesByUser() and added UserDetails as parameter. 
Changed user role condition to use userDetails.
`if (userDetails.getGrantedAuthorities().contains(ROLE_ROOT.name())`, `contains` does not work in any case, only text comparison of granted Authority text and Role name work.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java (hunk 2 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: Moved each block to a separate method.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java (hunk 3 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: Moved block to a separated method.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -104,48 +102,60 @@ public UserDTO addUserWithRoleAndApplication(UserRoleApplicationDTO userRoleAppl
...
+	private UserApplicationEntity createUserOrganization(UserEntity user, UserRoleApplicationDTO userRoleApplicationDTO) {
+		UserApplicationEntity userApplicationEntity = new UserApplicationEntity();
+		userApplicationEntity.setUser(user);
+		OrganizationEntity organization = organizationRepository.findById(userRoleApplicationDTO.getOrganization().getId());
+		userApplicationEntity.setOrganization(organization);
+		userApplicationEntity.setApplication(organization.getApplication());
+		return userApplicationRepository.save(userApplicationEntity);
+	}
+
+	private UserApplicationEntity createUserApplicationSameAsAdmin(UserEntity user, UserDetailsDTO userDetails) {
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: This method name is not reflecting what the method is doing. I believe the method is assigning a user to an application? I don't see it creating any userapplication. We need to improve this method name.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java [hunk 8]
```diff
@@ -74,24 +73,21 @@ public UserRoleDTO getUserRoleById(Long userRoleId) {
...
-			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_APP_ADMIN).build());
-			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_USER).build());
-		} else if (authentication.getAuthorities().toArray()[0].toString().equals(Role.ROLE_APP_ADMIN.name())) {
-			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_USER).build());
-			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_SURVEY_USER).build());
-//			rolesByLoggedUser.add(UserRoleDTO.builder().role(Role.ROLE_SOCIAL_ASSISTANT).build());
+	public List<UserRoleDTO> getAssignableRolesByUser(UserDetailsDTO userDetails) {
+		List<UserRoleDTO> getAssignableRolesByUser = new ArrayList<UserRoleDTO>();
+
+		if (userDetails.getAuthorities().toArray()[0].toString().equals(Role.ROLE_ROOT.name())) {
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: `userDetails.getAuthorities().contains(roleName)` did not work? 

`userDetails.getAuthorities().toArray()[0]` is assuming that the user will always have only one role assigned.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java (hunk 7 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: When an `ADMIN` user creates a simple user with role `ROLE_USER` or `ROLE_SURVEY_USER`, the user created belongs to the same Hub/Organization/Partner of the Admin user, it is not necessary to get that information from the frontend, that is why the User->Application/Organization relationship is copied from the admin (logged user). And yes, the method name can be improved, maybe `createUserApplicationForSimpleUser()`, any suggestion?

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java (hunk 8 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: `userDetails.getAuthorities().contains(roleName)` evaluates to `false` in all cases.
The following examples evaluates to `false` for a `Hub_Admin` user:
- userDetails.getAuthorities().contains(Role.ROLE_HUB_ADMIN)  -->  false
- userDetails.getAuthorities().contains(Role.ROLE_HUB_ADMIN.name())  -->  false
- userDetails.getAuthorities().contains(UserRoleDTO.builder().role(Role.ROLE_HUB_ADMIN).build())  -->  false

Comment (rodrigojv) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: I see. In that case, we can implement a new method with the `contains` logic.
```java
 private boolean userHasRole(UserDetaildDTO user, Role role)
    return user.getAuthorities()
      .stream()
       .map(a -> a.getAuthority())
      .collect(Collectors.toList())
      .contains(role.name();
```

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java (hunk 7 above)
Comment (rodrigojv) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: Some suggestions:
`addUserToAdminApplication(UserEntity newUser, UserDetaildDto adminUser)`
`addUserToApplication(UserEntity newUser, Long applicationId)`

File: src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java [hunk 9]
```diff
@@ -83,6 +100,34 @@ public OrganizationDTO getOrganizationById(Long organizationId) {
...
+	public List<OrganizationDTO> getOrganizationByLoggedUser(UserDetailsDTO userDetails) {
+
+		UserEntity user = userRepository.findOneByUsername(userDetails.getUsername()).orElseGet(UserEntity::new);
+		UserApplicationEntity userApplicationEntity = userApplicationRepository.findByUser(user).orElseGet(UserApplicationEntity::new);
+
+		OrganizationDTO organizationDTO = new OrganizationDTO();
+		if (userApplicationEntity.getOrganization() != null) {
+			organizationDTO = organizationMapper.entityToDto(userApplicationEntity.getOrganization());
+		} else {
+			organizationDTO.setName(userApplicationEntity.getApplication().getName());
```
[the remark this answers is not in the document]
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: Solved. Not necessary anymore. Removed in simplified User creation and Application/Organization assignments.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: Changed user role check, implemented with userHasRole() method, allowing multiple roles.

File: src/main/java/py/org/fundacionparaguaya/pspserver/web/rest/UserController.java (hunk 1 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/web/rest/UserController.java]: Commented code removed

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java (hunk 8 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserRoleServiceImpl.java]: Changed user role check, implemented userHasRole() method:
```
	private boolean userHasRole(UserDetailsDTO user, Role role) {
		return user.getAuthorities()
				.stream()
				.filter(auth -> auth.toString().equals(role.name()))
				.count() > 0;
	}
```

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java (hunk 7 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/services/impl/UserServiceImpl.java]: Method removed, not necessary in simplified User creation and Application/Organization assignments.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java [hunk 10]
```diff
@@ -1,38 +1,54 @@
...
+    private String username;
+
+    private String email;
+
+    private String pass;
 
     private Role role;
 
-    private ApplicationDTO application;
+    private Long application;
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java]: Here we can use `@JsonProperty("application")` and rename this property to `applicationId`. In that way the mapping will still work and we will have a more meaningful name for this property.

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java [hunk 11]
```diff
@@ -1,38 +1,54 @@
...
 
-    private OrganizationDTO organization;
+    private Long organization;
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java]: Use `@JsonProperty("organization")`

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java (hunk 10 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java]: Property `application` renamed to `applicationId`, mapped with `@JsonProperty`

File: src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java (hunk 11 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/security/dtos/UserRoleApplicationDTO.java]: Also, property `organization` renamed to `organizationId`, mapped with `@JsonProperty`

File: src/main/resources/db/changelog/sql/2018-01-10T09:50.sql [hunk 12]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE security.users
+  ADD COLUMN email character varying (100);
```
Comment (rodrigojv) [inline src/main/resources/db/changelog/sql/2018-01-10T09:50.sql]: En lugar de crear la columna cambiemos el tipo de dato de `text` a `character varying(256)`

File: src/main/resources/db/changelog/sql/2018-01-30T15:40.sql [hunk 13]
```diff
@@ -0,0 +1,4 @@
+--Change columns: name, code, description, application_id, is_active to NOT NULL
+ALTER TABLE security.users ALTER COLUMN email SET NOT NULL;
+ALTER TABLE security.users ALTER COLUMN pass SET NOT NULL;
+ALTER TABLE security.users ALTER COLUMN active SET NOT NULL;
```
Comment (rodrigojv) [inline src/main/resources/db/changelog/sql/2018-01-30T15:40.sql]: Cuidado con estas alteraciones. Agregar antes valores por defectos por si haya una columna con NULL.

```
update security.users set email = '' where email is NULL;
```

Lo mismo para las demás columnas.

File: src/main/resources/db/changelog/sql/2018-01-10T09:50.sql (hunk 12 above)
Comment (carlosbflores) [reply src/main/resources/db/changelog/sql/2018-01-10T09:50.sql]: Columna email cambiada a character varying (256)

File: src/main/resources/db/changelog/sql/2018-01-30T15:40.sql (hunk 13 above)
Comment (carlosbflores) [reply src/main/resources/db/changelog/sql/2018-01-30T15:40.sql]: Agregados valores por defecto a columnas antes de setear a NOT NULL

**Pull request merged**